addons/web/models/models.py

```diff
@@ -213,36 +219,200 @@ def fields_view_get(self, view_id=None, view_type='form', toolbar=False, submenu
 
+
+    @api.model
+    def search_panel_local_counters(self, field_name, **kwargs):
+        """
+        Return the record counts associated with the field_name values.
+
+        :param field_name: the name of a field (type many2one or selection)
+        :param category_domain: domain generated by categories
+        :param filter_domain: domain generated by filters
+        :param group_domain: domain generated by the field_name active values
+                                if field_name is used as a filter
+        :param search_domain: base domain of search
+        :return: a dict of the form {id: c,... } where id is a field_name value
+                    and c is the associated record count
+        """
+        count_domain = AND([
+            kwargs.get('search_domain', []),
+            kwargs.get('category_domain', []),
+            kwargs.get('filter_domain', []),
+            kwargs.get('group_domain', []),
+            [(field_name, '!=', False)]
+        ])
+
+        field = self._fields[field_name]
+
+        if field.type == 'many2one':
+            def get_group_id(group):
+                return group[field_name][0]
+
+        elif field.type == 'selection':
+            def get_group_id(group):
+                return group[field_name]
+
+        groups = self.read_group(count_domain, [field_name], [field_name])
+        local_counters = {}
+        for group in groups:
+            group_id = get_group_id(group)
+            local_counters[group_id] = group[field_name + '_count']
+
+        return local_counters
+
+
+    @api.model
+    def search_panel_global_counters(self, values_range, parent_name):
+        """
+        Modify in place values_range to transform the (local) counts
+        into global counts (local count + children local counts)
+        in case a parent field parent_name has been set on the range values.
+        Note that we save the initial (local) counts into an auxiliary dict
+        before they could be changed in the for loop below.
+
+        :param values_range: dict of the form
+            {
+                id: { 'id': id, 'count': c, parent_name: parent_id, ... }
+                ...
+            }
+        :param parent_name: string, indicates which key determines the parent
+        """
+        local_counters = lazymapping(lambda id: values_range[id]['count'])
+
+        for id in values_range:
+            values = values_range[id]
+            # here count is the initial value = local count set on values
+            count = local_counters[id]
+            if count:
+                parent_id = values[parent_name]
+                while parent_id:
+                    values = values_range[parent_id]
+                    local_counters[parent_id]
+                    values['count'] += count
+                    parent_id = values[parent_name]
+
+
     @api.model
-    def search_panel_select_range(self, field_name):
+    def search_panel_range(self, field_name, **kwargs):
```

Comment:
This is the refactoring, shouldn't this method be private?

Reply:
ok